packer/registry_builder.go

```diff
@@ -0,0 +1,106 @@
+package packer
+
+import (
+	"context"
+	"log"
+	"strings"
+
+	"github.com/hashicorp/hcp-sdk-go/clients/cloud-packer-service/preview/2021-04-30/models"
+	packersdk "github.com/hashicorp/packer-plugin-sdk/packer"
+	packerregistry "github.com/hashicorp/packer/internal/packer_registry"
+)
+
+type RegistryBuilder struct {
+	Name                      string
+	ArtifactMetadataPublisher *packerregistry.Bucket
+
+	packersdk.Builder
+}
+
+func (b *RegistryBuilder) Prepare(raws ...interface{}) ([]string, []string, error) {
+	return b.Builder.Prepare(raws...)
+}
+
+// Run is where the actual build should take place. It takes a Build and a Ui.
+func (b *RegistryBuilder) Run(ctx context.Context, ui packersdk.Ui, hook packersdk.Hook) (packersdk.Artifact, error) {
+	runCompleted := make(chan struct{})
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				log.Printf("[TRACE] marking build %q as cancelled in Packer registry", b.Name)
+				if err := b.ArtifactMetadataPublisher.PublishBuildStatus(context.TODO(), b.Name, models.HashicorpCloudPackerBuildStatusCANCELLED); err != nil {
+					log.Printf("[TRACE] failed to update Packer registry status for %q: %s", b.Name, err)
+				}
+				return
+			case <-runCompleted:
+				return
+			}
+		}
+	}()
+
+	if err := b.ArtifactMetadataPublisher.PublishBuildStatus(ctx, b.Name, models.HashicorpCloudPackerBuildStatusRUNNING); err != nil {
+		log.Printf("[TRACE] failed to update Packer registry status for %q: %s", b.Name, err)
+	}
+
+	artifact, err := b.Builder.Run(ctx, ui, hook)
+	if err != nil {
+		if parErr := b.ArtifactMetadataPublisher.PublishBuildStatus(ctx, b.Name, models.HashicorpCloudPackerBuildStatusFAILED); parErr != nil {
+			log.Printf("[TRACE] failed to update Packer registry status for %q: %s", b.Name, parErr)
+		}
+	}
+
+	// close chan to mark completion
+	close(runCompleted)
+
+	var artifacts []packersdk.Artifact
+	artifacts = append(artifacts, artifact)
+
+	for _, artifact := range artifacts {
+		// Lets post state
+		if artifact != nil {
+			metadata := make(map[string]string)
+			metadata[artifact.BuilderId()] = artifact.String()
+			metadata[artifact.BuilderId()+".files"] = strings.Join(artifact.Files(), ", ")
+			err := b.ArtifactMetadataPublisher.AddBuildMetadata(b.Name, metadata)
+			if err != nil {
+				log.Printf("[TRACE] failed to add build labels for %q: %s", b.Name, err)
+			}
+
+			switch state := artifact.State("par.artifact.metadata").(type) {
+			case map[interface{}]interface{}:
+				m := make(map[string]string)
+				for k, v := range state {
+					m[k.(string)] = v.(string)
```

Comment:
if v is for some reason not a string this will panic

Reply:
Is it possible not to be a string?

Reply:
So right it is hard coded, but yes it could as the metadata can be set by the maintainer right now. We need to think how to properly send the data over rpc and ensure that it has the types that we expect. I can see this being an issue when we start to send the Metadata in the response which is of type `map[string]string`